### DISCUSSION TITLE
How to listen to StreamProvider.autoDispose inside StateNotifierProvider?

### DISCUSSION THREAD

**Discussion by ebelevics** [rrousselGit/riverpod#1045] [Q&A]
I have WebSocket which is tied to `final wsProvider = StreamProvider.autoDispose<WSMessage>`. Then I have `authProvider = StateNotifierProvider` which is not autodisposed, because I need to listen to user authentication state all the time. 

I have come to problem that I can't listen to wsProvider inside of authProvider's StateNotifier with Ref because of ``The argument type 'AutoDisposeStreamProvider<WSMessage>' can't be assigned to the parameter type 'AlwaysAliveProviderListenable<Object>'.` 

I also can't use AutoDisposeRef because both providers have different dispose logic. Only by changing wsProvider to normal SteamProvider will remove the error, but this is not expected usecase, as WS should dispose and reinstantiate for each a user during login/logout. Also because wsProvider is listened in other autodisposed providers.

How can I solve this issue?

**Comment by rrousselGit** [comment]
> I also can't use AutoDisposeRef because both providers have different dispose logic. 

I don't understand what this means. You should be able to make your statenotifierprovider use autoDispose without issue

**Reply by ebelevics** [reply]
```dart
final messageProvider = StreamProvider.autoDispose<String>((ref) async* {
  final channel = StreamController.broadcast();

  ref.onDispose(() => channel.sink.close());

  await for (final value in channel.stream) {
    yield value.toString();
  }
});

final authProvider = StateNotifierProvider<AuthNotifier, String>((ref) {
  return AuthNotifier(ref);
});

class AuthNotifier extends StateNotifier<String> {
  Ref ref;

  AuthNotifier(this.ref) : super("") {
    ref.listen(messageProvider, (previous, next) {}); //error here
  }
}

final otherProvider = StateNotifierProvider.autoDispose<OtherNotifier, String>((ref) {
  return OtherNotifier(ref);
});

class OtherNotifier extends StateNotifier<String> {
  AutoDisposeRef ref;

  OtherNotifier(this.ref) : super("") {
    ref.listen(messageProvider, (previous, next) {});
  }
}
```
This is what I meant, I can't use ref.listen as I get already mentioned error, if I want to use in two providers one normal other autodispose

**Comment by iamarnas** [comment]
@ebelevics You should change your `StateNotifierProvider` to auto dispose. Is nothing wrong to use auto dispose, you can make all your providers auto dispose. Auto dispose methods are created to warn user about mixing them and have some extra stuff.

**Reply by ebelevics** [reply]
Yeah, didn't thought about it, but it seems as only solution for now. Need to make sure that I watch it at start of app, so it doesn't auto dispose.

**Reply by iamarnas** [reply]
> Yeah, didn't thought about it, but it seems as only solution for now. Need to make sure that I watch it at start of app, so it doesn't auto dispose.

This is a right solution. If choose combine providers with Riverpod then you need choice auto disposed combined providers or no.

**Reply by ebelevics** [reply]
Ok, thank you

**Comment by ebelevics** [comment]
Answered
